### PR TITLE
SB-299 making sure maven-metadata.xml is up-to-date

### PR DESCRIPTION
**Pull request by fuss86**

### CODE REVIEW

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/MavenGroupRepositoryComponent.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.lang3.mutable.MutableBoolean;
+import org.apache.maven.artifact.repository.metadata.Metadata;
+import org.springframework.stereotype.Component;
+
+/**
+ * @author Przemyslaw Fusik
+ */
+@Component
+public class MavenGroupRepositoryComponent
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/MavenGroupRepositoryComponent.java]: I think a better name would be `MavenMetadataGroupRepositoryComponent`. Also, these should probably be in a sub-package called `metadata`.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/MavenGroupRepositoryComponent.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    @Inject
+    private LayoutProviderRegistry layoutProviderRegistry;
+
+    @Inject
+    private ArtifactMetadataService artifactMetadataService;
+
+    @Inject
+    private MavenMetadataManager mavenMetadataManager;
+
+    public void deleteMetadataInRepositoryParents(final String storageId,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/MavenGroupRepositoryComponent.java]: Parents? I think the name is a bit dubious... How about `deleteMetadataInGroupRepositoriesContainingPath`...?

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/MavenGroupRepositoryComponent.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    /**
+     * This method goes down the group repository tree and lookups for an existence of the artifact in the given path.
+     */
+    private void checkArtifactExistenceInTheGroupRepositorySubtree(final Repository groupRepository,
+                                                                   final String artifactRelativePath,
+                                                                   final Map<String, MutableBoolean> repositoryArtifactExistence,
+                                                                   final MutableBoolean artifactExistence)
+            throws IOException
+    {
+        for (final String maybeStorageAndRepositoryId : groupRepository.getGroupRepositories())
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/MavenGroupRepositoryComponent.java]: Haha! I love the name of this variable!... :D

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/MavenGroupRepositoryComponent.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        {
+            final String subStorageId = configurationManager.getStorageId(groupRepository.getStorage(),
+                                                                          maybeStorageAndRepositoryId);
+            final String subRepositoryId = configurationManager.getRepositoryId(maybeStorageAndRepositoryId);
+            final Repository subRepository = configurationManager.getConfiguration().getStorage(
+                    subStorageId).getRepository(
+                    subRepositoryId);
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/repository/MavenGroupRepositoryComponent.java]: Cosmetic: If you could just re-indent it, please:

            final Repository subRepository = configurationManager.getConfiguration()
                                                                 .getStorage(subStorageId)
                                                                 .getRepository(subRepositoryId);

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/storage/metadata/MavenMetadataManager.java [hunk 5]
```diff
@@ -348,23 +325,50 @@ public Metadata generateSnapshotVersioningMetadata(Path snapshotBasePath,
...
+            {
+                try
+                {
+                    final Metadata metadata = readMetadata(metadataBasePath);
+                    mergeAndStore(metadataBasePath, metadata, mergeMetadata);
+                    return;
+                }
+                catch (Exception e)
+                {
+                    logger.error("Unable to merge the metadata to " + metadataBasePath + " by source metadata" +
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/storage/metadata/MavenMetadataManager.java]: Cosmetic: It looks like there's a missing space in `" by source metadata"` and no dot before `" Continuing..."`?

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/repository/MavenGroupRepositoryComponentTest.java [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+import javax.xml.bind.JAXBException;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.Random;
+import java.util.Set;
+
+import edu.emory.mathcs.backport.java.util.Arrays;
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/repository/MavenGroupRepositoryComponentTest.java]: Cosmetic: This looks like an incorrect import.

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/repository/MavenGroupRepositoryComponentTest.java [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+        // grand parent with other kids
+        metadata = mavenMetadataManager.readMetadata(
+                layoutProvider.resolve(createRepositoryMock(STORAGE0, REPOSITORY_GROUP_A)).resolve(
+                        "com/artifacts/to/delete/releases/delete-group"));
+        assertThat(metadata.getVersioning().getVersions().size(), CoreMatchers.equalTo(2));
+        assertThat(metadata.getVersioning().getVersions().get(0), CoreMatchers.equalTo("1.2.1"));
+        assertThat(metadata.getVersioning().getVersions().get(1), CoreMatchers.equalTo("1.2.2"));
+    }
+
+    /*
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/repository/MavenGroupRepositoryComponentTest.java]: Is this work-in-progress (and is therefore commented out), or is it something that will be removed?

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/event/artifact/BaseMavenArtifactEventListener.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    @Inject
+    MavenGroupRepositoryComponent mavenGroupRepositoryComponent;
+
+    Repository getRepository(final ArtifactEvent event)
+    {
+        return configurationManager.getConfiguration().getStorage(event.getStorageId()).getRepository(
+                event.getRepositoryId());
+    }
+
+    void updateParentGroupRepositoriesMetadata(final ArtifactEvent event,
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/main/java/org/carlspring/strongbox/event/artifact/BaseMavenArtifactEventListener.java]: Perhaps `updateMetadataInGroupsContainingPath` would be a better name? Although, to be correct, I guess it should be "groups containing repository"...?

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/repository/MavenGroupRepositoryComponentTest.java (hunk 7 above)
Comment (fuss86) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-maven-layout-provider/src/test/java/org/carlspring/strongbox/repository/MavenGroupRepositoryComponentTest.java]: garbage :)

**Comment by fuss86** [comment]
@carlspring, this PR is ready to be merged.

**Pull request merged**